allow offers request on orderbook subscribe when disconnected

shekenahglory opened the pull request:
When the remote is offline and an orderbook is created with XRP as the `gets` currency, the orderbook never gets properly subscribed because requestOffers doesn't allow a request to go to the queue when remote is disconnected.  Its not a problem for non-native currencies because requestTransferRate waits for the remote to connect.

FrRichard commented:
is it related to https://github.com/ripple/ripple-lib/issues/573 ?

shekenahglory commented:
Seems likely - if you are subscribing before connecting

FrRichard commented:
hmm this is weird because I always remote.connect() before .. I'll try with those changes and see

FrRichard commented:
this bug didn't appear with 0.12.6

shekenahglory commented:
remote.connect() is async, so it may be that you initiated the connection but hasnt been completed yet - unless you are specifically waiting until you get a connection response

FrRichard commented:
got it thanks

clark800 commented:
Is it possible to add a unit test to prevent regressions? Could you add a comment explaining what `internal` is and how it is used?

shekenahglory commented:
Yeah I can add that, I'm not entirely sure that the bit about ignoring requests while its not connected is ideal though...it doesn't seem like expected behavior does it?

shekenahglory commented:
It seems like we would want to ignore subsequent requests if one is already queued up, not ignore all

clark800 commented:
I think the Request class will call this method again when it reconnects, until it gets a success or error response or times out.

shekenahglory commented:
As it is now, if requestOffers is called when its disconnected, a request object never gets created - it just returns undefined without even executing the callback

shekenahglory commented:
Actually isnt this always an internal call and therefor a private method? there is another method `getOffers` that is public

shekenahglory commented:
@darkdarkdragon I also just discovered that autobridged books don't load because `syncronized` is never set to `true`: https://github.com/ripple/ripple-lib/blob/develop/src/core/orderbook.js#L358 I thinking this supposed to be `synced` here and under `computeAutobridgedOffersWrapper` also?

darkdarkdragon commented:
@shekenahglory You're right, it must be `_synced`. Missed that one on merge :confused:

clark800 commented on src/core/orderbook.js:
```diff
@@ -333,16 +336,18 @@ OrderBook.prototype.destroy = function() {
     // do not make request if not online.
     // that requests will be queued and
     // eventually all of them will fire back
+    callback(new Error('remote is offline'));
```
All errors passed to callbacks should be instances of RippleError (which means they can inherit from RippleError).

clark800 commented:
LGTM after squash

shekenahglory commented:
Rebased and squashed, but now I get seemingly unrelated test failures?

`ReferenceError: UInt160 is not defined
      at Remote.requestGatewayBalances (/home/ubuntu/ripple-lib/src/core/remote.js:1:14)`

clark800 commented:
Yes, that was a merge failure that happened yesterday. I'll fix it soon, thanks

clark800 commented:
I fixed the bug and tried rebuilding on CircleCI, but it failed again. Can you try rebasing?
